### PR TITLE
MINOR: Fix ProducerPerformance still counting successful sending when sending failed

### PR DESCRIPTION
**Pull request by hudeqi**
When using ProducerPerformance, I found that when the sending fails, it is still counted as successfully sent by stat and the metrics are printed. For example, when there is no write permission and cannot be written in, the sending success rate is still magically displayed.

### CODE REVIEW

**Comment by hudeqi** [comment]
Can someone take a look? @mimaison @chia7712

**Review by showuon — COMMENTED** [review]
@hudeqi , thanks for the patch. This change makes sense to me. Could you add tests for this change?

**Comment by hudeqi** [comment]
>

**Comment by hudeqi** [comment]
> >

**Comment by hudeqi** [comment]
> @hudeqi , thanks for the patch. This change makes sense to me. Could you add tests for this change?

Do you want to add test screenshots/results for this case, or the test code corresponding to this change? @showuon

**Comment by showuon** [comment]
The test code, I mean. We already have some unit tests for producerPerformance [here](https://github.com/apache/kafka/blob/trunk/tools/src/test/java/org/apache/kafka/tools/ProducerPerformanceTest.java). Thanks.

**Comment by hudeqi** [comment]
> 

Hello, the relevant test code has been submitted, please help to review, thank you. @showuon

File: tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java [hunk 1]
```diff
@@ -113,7 +113,7 @@ void start(String[] args) throws IOException {
                 record = new ProducerRecord<>(topicName, payload);
 
                 long sendStartMs = System.currentTimeMillis();
-                Callback cb = stats.nextCompletion(sendStartMs, payload.length, stats);
+                cb = stats.nextCompletion(sendStartMs, payload.length, stats);
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java]: Does it need to pass itself (`stats`) to nextCompletion? It seems to me line#399 could be replaced by `new PerfCallback(this.iteration, start, bytes, this);`

**Review by chia7712 — COMMENTED** [review]
@hudeqi thanks for contribution. one minor comment left. PTAL

File: tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java (hunk 1 above)
Comment (hudeqi) [reply tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java]: A great and careful code optimization.

**Review by showuon — APPROVED** [review]
LGTM!

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by hudeqi** [comment]
Thx, Is this [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-842%3A+Add+richer+group+offset+reset+mechanisms) interested in understanding and discussing? @chia7712 @showuon

**Comment by robobario** [comment]
Hi, I think this introduced a race condition that can cause the producer-perf-test to log a lot of exceptions to console if num records > 500000. Since the `iteration` is incremented from the callback, multiple threads may see the same value when deciding whether to update `latencies`, so it can go out of bounds when updating `latencies`.
```
bin/kafka-producer-perf-test.sh     --topic perf-test     --throughput -1     --num-records 1000000     --producer-props acks=all bootstrap.servers=localhost:9092     --record-size 50
```
Sometimes it logs:
```
[2023-03-16 15:30:57,432] ERROR Error executing user-provided callback on message for topic-partition 'perf-test-0' (org.apache.kafka.clients.producer.internals.ProducerBatch)
java.lang.ArrayIndexOutOfBoundsException: Index 500001 out of bounds for length 500001
	at org.apache.kafka.tools.ProducerPerformance$Stats.record(ProducerPerformance.java:38
```
It doesn't kill the test run but it looks alarming to see a couple of hundred errors in console.

**Comment by showuon** [comment]
@robobario , nice catch! Welcome to submit a PR to fix it. Thank you.

**Comment by hudeqi** [comment]
> Hi, I think this introduced a race condition that can cause the producer-perf-test to log a lot of exceptions to console if num records > 500000. Since the `iteration` is incremented from the callback, multiple threads may see the same value when deciding whether to update `latencies`, so it can go out of bounds when updating `latencies`.
> 
> ```
> bin/kafka-producer-perf-test.sh     --topic perf-test     --throughput -1     --num-records 1000000     --producer-props acks=all bootstrap.servers=localhost:9092     --record-size 50
> ```
> 
> Sometimes it logs:
> 
> ```
> [2023-03-16 15:30:57,432] ERROR Error executing user-provided callback on message for topic-partition 'perf-test-0' (org.apache.kafka.clients.producer.internals.ProducerBatch)
> java.lang.ArrayIndexOutOfBoundsException: Index 500001 out of bounds for length 500001
> 	at org.apache.kafka.tools.ProducerPerformance$Stats.record(ProducerPerformance.java:38
> ```
> 
> It doesn't kill the test run but it looks alarming to see a couple of hundred errors in console.

 nice catch!

**Comment by chia7712** [comment]
@robobario thanks for feedback. According to your nice catch `Stats` is shared between all producer thread. Not only `iteration` but also other mutable variables should be thread-safe, or we may see the incorrect stats value, right?

**Comment by hudeqi** [comment]
> 

In fact, I also have this question. The variables in stats are all thread-unsafe at present, but the printed data looks fine😂, and this is a probabilistic problem for the version I repaired.

**Comment by robobario** [comment]
Maybe my thread safety explanation is incorrect, I'll look again tomorrow.

**Comment by chia7712** [comment]
> Maybe my thread safety explanation is incorrect, I'll look again tomorrow.

the thread safety explanation is correct to me. My point was that we should not only fix `iteration` but also fix other mutable stat values.

**Comment by hudeqi** [comment]
> > Maybe my thread safety explanation is incorrect, I'll look again tomorrow.
> 
> the thread safety explanation is correct to me. My point was that we should not only fix `iteration` but also fix other mutable stat values.

If all variables are volatile or atomic, will it affect the printing performance of this perf program?

**Comment by chia7712** [comment]
> If all variables are volatile or atomic, will it affect the printing performance of this perf program?

Or we just handle the `iteration` as it seems `iteration` is the only variable used by main thread and producer thread at the same time.

**Comment by hudeqi** [comment]
> 



> > If all variables are volatile or atomic, will it affect the printing performance of this perf program?
> 
> Or we just handle the `iteration` as it seems `iteration` is the only variable used by main thread and producer thread at the same time.

I understand what you mean, because this is a separate perf producer program, there should be no need to consider the scenario of multiple producer threads, because this perf program is sent by one producer, but this problem is indeed caused by the competition between the main thread and the producer thread , so only this variable needs to be fixed.
In addition, does this repair need to open a new jira?

**Comment by chia7712** [comment]
> In addition, does this repair need to open a new jira?

In fact, we should revert this PR first. The solution offered by this PR bring antoher issue (thanks to @robobario). The origin issue is worth having a JIRA since it is related to a true bug.

**Comment by hudeqi** [comment]
> > In addition, does this repair need to open a new jira?
> 
> In fact, we should revert this PR first. The solution offered by this PR bring antoher issue (thanks to @robobario). The origin issue is worth having a JIRA since it is related to a true bug.

Ok, that is to say, I need to create a new jira, and then report back the original problem that failed to send but still recorded success, and solve the current problem at the same time, right?

**Comment by chia7712** [comment]
> and then report back the original problem that failed to send but still recorded success, and solve the current problem at the same time, right?

yep. #13400 is used to revert this change.

**Pull request merged**